### PR TITLE
PLT-7500: Adds emoji skintone support.

### PR DESCRIPTION
**Pull request by mkraft**
#### Summary
This features replaces emoji that have a skin tone modifier with their corresponding png image if the following conditions are met:

a) The emoji modifier base is among the db of emoji in the [gemoji ruby gem](https://github.com/github/gemoji).
b) The emoji is a member of the set of emoji modifier bases as listed by Unicode.
c) The corresponding image is among the Apple Color Emoji system font set.

The `make emojis` command was used to add all of the included new assets.

#### Ticket Link
[Jira PLT-7500](https://mattermost.atlassian.net/browse/PLT-7500)
[Github issue 7337](https://github.com/mattermost/mattermost-server/issues/7337)

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes

### CODE REVIEW

**Comment by mkraft** [comment]
screen shot 2017-10-20 at 9 42 57 pm
screen shot 2017-10-20 at 9 43 11 pm
screen shot 2017-10-20 at 9 43 22 pm

**Comment by jasonblais** [comment]
Awesome thanks @mkraft! 

@hmhealey is probably the best person to share initial feedback, but I think we don't actually want to show the alternate skin tones in emoji picker or emoji autocomplete.

**Comment by ccbrown** [comment]
Yeah, the GitHub issue actually says specifically:

> Don't show the alternative skin tones in emoji picker or emoji autocomplete

I think the scope of this issue was just to include images so that they display in messages consistently.

When it does become time to add a UI for selecting skin tones, I suspect we'll want to make it a two-phase process since all the variants really add a lot of noise to the searching / browsing process. For example...

Slack:

screen shot 2017-10-21 at 12 02 30 am

Apple:

screen shot 2017-10-21 at 12 04 11 am

But yeah, any picker or autocomplete changes seem to be beyond the scope of this issue.

**Comment by mkraft** [comment]
@ccbrown Yes I saw that in the story but decided to leave the new emojis in the auto suggest and picker as a talking point before doing the added effort to remove them. Expect a commit with the removals.

**Comment by mkraft** [comment]
Note that it looks like we'll need to do one of the following:

a) Pay for an [EmojiOne license](https://www.emojione.com/licenses) to use their emoji attribution-free.
b) Place attributions according to their terms.
c) Choose another download source for the new Emoji.

**Comment by ccbrown** [comment]
@mkraft 👍

Did you intend to close this PR?

**Comment by mkraft** [comment]
@ccbrown Yes. New to your process. Would you normally keep this open or open a new one? Is it OK to squash and force push an open PR?

**Comment by ccbrown** [comment]
@mkraft It's okay to just push additional commits, amend commits, squash commits, anything you like. Force pushing to a PR branch isn't going to trip us up. And in the end, we're just gonna squash any changes you submit to the PR as part of the merging process anyways.

So usually we just keep one PR open, and iterate on it.

**Comment by mkraft** [comment]
@ccbrown Perfect that's the best of both worlds: clean commit history and not too many PRs.

**Comment by mkraft** [comment]
Re-opened it by doing https://gist.github.com/robertpainsi/2c42c15f1ce6dab03a0675348edd4e2c. Lol.

File: components/suggestion/emoticon_provider.jsx [hunk 1]
```diff
@@ -81,7 +82,7 @@ export default class EmoticonProvider {
                 if (emoji.aliases) {
                     // This is a system emoji so it may have multiple names
                     for (const alias of emoji.aliases) {
-                        if (alias.indexOf(partialName) !== -1) {
+                        if (!EMOJI_CATEGORY_SUGGESTION_BLACKLIST.includes(emoji.category) && alias.indexOf(partialName) !== -1) {
```
Comment (mkraft) [inline components/suggestion/emoticon_provider.jsx]: I could use some opinions about whether there is a better place to be filtering-out these suggestions.

File: make-emojis [hunk 2]
```diff
@@ -37,6 +30,37 @@ custom_aliases.each do |custom, original|
     end
 end
 
+# @param file_name [String] a PNG filename, ex: '1f4aa-1f3ff.png'
+# @return [String] the PNG bytes
+def download_emoji_png(file_name)
+    source_url = "https://github.com/EmojiTwo/emojitwo/raw/master/png/#{file_name.downcase}"
```
Comment (mkraft) [inline make-emojis]: I'm pulling new emoji from EmojiTwo which is a clone of the last set of free emoji from EmojiOne. So we should be good from a license point of view.

**Comment by mkraft** [comment]
I have removed the typeahead suggestions and picker options for skin tones.

File: components/emoji_picker/emoji_picker.jsx [hunk 3]
```diff
@@ -115,7 +115,7 @@ export default class EmojiPicker extends React.Component {
                 name: 'flags',
                 className: 'fa fa-flag-o',
                 id: 'emoji_picker.flags',
-                message: 'Fags',
+                message: 'Flags',
```
Comment (mkraft) [inline components/emoji_picker/emoji_picker.jsx]: Noticed a typo unrelated to this story.

Comment (ccbrown) [reply components/emoji_picker/emoji_picker.jsx]: That's quite a typo. Fortunately that gets overridden by translations and isn't actually displayed anywhere under normal circumstances. :-/

**Comment by mkraft** [comment]
@ccbrown Regarding the flipping of the first/last image, just based on my 2-minute investigation, `SpriteFactory.run!` takes a directory of images and then places them somehow (we don't really care how) and then returns the position of each image (etc). Since the file names we're providing aren't different I presume the change is in either the `sprite_factory` or `chunky_png` gem.

**Comment by jasonblais** [comment]
Tested posting emojis with all available skin tones on iOS 10 / iPhone 5s / Mattermost mobile app. Works as expected, including after refresh and clearing cache.

image

You can post the alternate skin tones via `:wave_skin_tone_3:`, but they do not show in the emoji picker or the emoji autocomplete, as expected.

image

Moving to dev review for further reviews.

**Comment by ccbrown** [comment]
Code changes look fine to me, but the emojis we download from emojitwo are a very different style than our others. I would expect this:

screen shot 2017-10-22 at 2 00 46 pm

To look like this:

screen shot 2017-10-22 at 2 04 03 pm

It's not too hard to make Gemoji output all of the skin-tone emojis, but you would need to modify a few lines of the extractor (see https://github.com/github/gemoji/pull/101). I think we should pursue that route so all of the emojis look more consistent.

**Comment by jasonblais** [comment]
@ccbrown Aah good catch actually. Agree

File: components/emoji_picker/emoji_picker.jsx (hunk 3 above)
Comment (saturninoabril) [reply components/emoji_picker/emoji_picker.jsx]: Nice catch!

**Comment by mkraft** [comment]
@ccbrown That was the first route I pursued and had that solution working originally but abandoned it because it became evident that there are a lot of emoji that are not included in that 1-year-old PR. 

In other words the JSON DB is out of sync with all of the latest emoji in gemoji and it looks like Github won't be merging that change in. 

So we can:

a) create our own fork of gemoji and try to merge our custom DB with Github's DB every time they include new emoji. Note that the emoji downloader wasn't included by the author of the PR to gemoji.
b) do what I did.

I suggest we simply point `download_emoji_png` to a different source that has the same style of emoji as our current set.

**Comment by mkraft** [comment]
Implemented my suggestion with the matching emoji to our original set.

**Comment by saturninoabril** [comment]
@mkraft Could you rebase to latest master?

**Comment by mkraft** [comment]
@saturninoabril Done.

**Comment by ccbrown** [comment]
@mkraft 

> Regarding making your suggested change to gemoji that was the first solution I pursued (and had that working) but abandoned it because it became evident that there are a lot of emoji that are not included in that 1-year-old PR.
> 
> In other words the emoji.json file is out of sync with all of the latest emoji that Github has added and it doesn't look like Github will be merging that change in.

Yeah I wasn't proposing that we just build using that PR's fork/branch. I was proposing something more along the lines of copying extractor.rb to the webapp repo and modifying it to export skin tones. It'd have the upsides of not requiring an internet connection, probably being faster, and possibly simpler since all of the emojis would be obtained in the same way.

**Comment by mkraft** [comment]
> I was proposing something more along the lines of copying extractor.rb to the webapp repo and modifying it to export skin tones (I don't think the db changes made in that PR are actually required for that.). 

Would that actually bring the new emoji PNGs into our images directory? I could be wrong but at quick look it seems like the PNGs need to be in your images directory already for `gemoji` to use them.

> It'd have the upsides of not requiring an internet connection, probably being faster, and possibly being simpler since all of the emojis would be obtained in the same way.

I'm not too concerned about an internet connection required for the occasional pulling of new emoji. If we're pulling new emoji we're likely updating the `gemoji` gem anyhow which would also use an internet connection. Also we can still manually add emoji to the images directory and the download method just skips the file.

Correct me if I am mistaken on either of those points because maybe I misinterpreted something.

**Comment by ccbrown** [comment]
@mkraft 

Gemoji extracts the emojis from your system font. You don't need to have any images beforehand.

The internet connection point is pretty minor. Not really much of a concern if any; just trying to weigh the pros / cons.

**Comment by mkraft** [comment]
@ccbrown 

> Gemoji extracts the emojis from your system font. You don't need to have any images beforehand.

You're right now I remember because `db:generate` doesn't run on my MBP because of [this bug](https://github.com/github/gemoji/issues/133). 

But regardless of that, it's trivial for me to switch to pull the PNGs from my system font if there's an appetite for that? In the end it might be the same bytes, but yes it would be faster and from the same source (depending on OS updates, of course).

Thoughts?

**Comment by ccbrown** [comment]
@mkraft If it's trivial, I do think we should switch. Reducing dependencies in general is good.

**Comment by mkraft** [comment]
> it's trivial for me to switch to pull the PNGs from my system font

... I said with no knowledge of how trivial it is. No worries, I'll take a look tonight: best to do it right the first time.

**Comment by mkraft** [comment]
@ccbrown I have changed this PR to extract the emoji PNGs from the Apple Color Emoji Emoji system font as discussed.

**Comment by ccbrown** [comment]
Awesome. Looks good to me, but there are now like 7 top-level files used in the emoji generation. I think moving those into a subdirectory is probably warranted now. Other than that, looks good to me! And you'll need to rebase due to the age of this PR.

**Comment by mkraft** [comment]
@ccbrown You are correct that is a lot of files! I have moved all 8 emoji-related build files into `build/emoji`.

**Comment by ccbrown** [comment]
Excellent. Waiting on @hmhealey then. And you'll still need to rebase this due to its age.

**Comment by mkraft** [comment]
@ccbrown Done. Lemme know if it worked as the forked-repo-rebase flow is new to me.

**Comment by jasonblais** [comment]
Yeah, let's wait until v4.4 branch is cut on Monday and merge this PR to `master`.

Big thanks @mkraft :+1:

**Pull request merged**